### PR TITLE
Feat (brevitas_examples/llm): better dataset handling

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR

Dataset management is not ideal, with iterator being explicitly converted to lists for no apparent reason.

This should take care of that, although it would be worth taking a more general look in the future.

We need to add 'labels' token everywhere, to avoid a potential miss-match between the forward signature in training/eval mode after the model has gone through FX tracing.

### CODE REVIEW

File: src/brevitas_examples/llm/llm_quant/data.py
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data.py]:
```suggestion
) -> datasets.Dataset:
```

File: src/brevitas_examples/llm/llm_quant/data.py [hunk 1]
```diff
@@ -179,29 +172,27 @@ def get_wikitext2(
     # Add BOS token to each sequence if add_bos_token is True and the tokenizer supports this token
     if add_bos_token and tokenizer.bos_token_id is not None:
         seqlen = seqlen - 1
-        sequence_process_fn = lambda inp: torch.cat([
-            torch.tensor([[tokenizer.bos_token_id]], dtype=inp.dtype, device=inp.device), inp],
-                                                    dim=1)
+        sequence_process_fn = sequence_process_fn = lambda inp: [tokenizer.bos_token_id] + inp
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data.py]: Should this be?
```suggestion
        sequence_process_fn = lambda inp: [tokenizer.bos_token_id] + inp
```

File: src/brevitas_examples/llm/llm_quant/data.py
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data.py]:
```suggestion
        seed: int = 42) -> datasets.Dataset:
```

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 2]
```diff
@@ -44,22 +45,59 @@
 from brevitas_examples.llm.llm_quant.data import load_raw_dataset
 
 
-class DatasetToDevice(torch.utils.data.Dataset):
+def collate_batch(examples: List[Dict[str, List[np.ndarray]]],
+                  sequence_length) -> Dict[str, torch.Tensor]:
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py]:
```suggestion
                  sequence_length: int) -> Dict[str, torch.Tensor]:
```

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 3]
```diff
@@ -44,22 +45,59 @@
...
+    # Make sure we load only what's necessary, ie we only load a `input_ids` column.
+    assert all(list(example.keys()) == ["input_ids"] for example in examples)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py]: Should we remove this? Maybe it is worth turning into an `if ... raise ValueError(....)`.

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 4]
```diff
@@ -44,22 +45,59 @@
...
 
-    def __init__(self, data: List, device: Optional[Union[str, torch.device]]):
-        super().__init__()
-        self.data = data
-        self.device = device
+    input_ids = np.vstack([examples[i]["input_ids"] for i in range(len(examples))])  # (b, s)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py:54]: This is a `numpy.ndarray` still? Shouldn't it be converted to `torch.Tensor`? I think `torch.vstack` should do the trick.

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 5]
```diff
@@ -44,22 +45,59 @@
...
 
-    def __getitem__(self, idx):
-        if self.device is not None:
-            return {
-                name: recursive_to_device(val, self.device) for name, val in self.data[idx].items()}
-        else:
-            return self.data[idx]
+    result: Dict[str, np.ndarray] = {}
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py]:
```suggestion
    result: Dict[str, torch.Tensor] = {}
```

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 6]
```diff
@@ -44,22 +45,59 @@
...
-    def __len__(self):
-        return len(self.data)
+        num_heads = normalized_config.num_attention_heads
+        if hasattr(normalized_config, "num_key_value_heads"):
+            num_kv_heads = normalized_config.num_key_value_heads
+        else:
+            num_kv_heads = num_heads
+        head_dim = normalized_config.hidden_size // num_heads
+        return partial(collate_batch_fx, num_kv_heads=num_kv_heads, head_dim=head_dim)
+    else:
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py]: No need to have `else`, as the `if` has a return statement. This could just be:
```
if require_fx:
     return partial(...)
return partial(...)
```

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 7]
```diff
@@ -81,19 +119,17 @@ def recursive_to_device(tensor_or_iterable: Union[Iterable, torch.Tensor], devic
...
+        dataset_name: str,
+        tokenizer: Any,
+        nsamples: int = 128,
+        seqlen: int = 2048,
+        seed: int = 0,
+        split: str = "train",
+        bos_preprocessing: Optional[str] = None,
+        add_eos_token: bool = False,
+        fuse_documents: bool = True,
+        require_fx: bool = False):
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py]:
```suggestion
        require_fx: bool = False) -> datasets.Dataset:
```

**Review by pablomlago — CHANGES_REQUESTED** [review]
Some minor comments

File: src/brevitas_examples/llm/llm_quant/data_utils.py (hunk 4 above)
Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_quant/data_utils.py:54]: I added a conversion to torch.tensor after the `np.vstack`. Unless torch.vstack is orders of magnitude faster than np.vstack, I'd say it's fine

File: src/brevitas_examples/llm/llm_quant/data_utils.py [hunk 8]
```diff
@@ -81,19 +118,15 @@ def recursive_to_device(tensor_or_iterable: Union[Iterable, torch.Tensor], devic
...
-):
+        dataset_name: str,
+        tokenizer: Any,
+        nsamples: int = 128,
+        seqlen: int = 2048,
+        seed: int = 0,
+        split: str = "train",
+        bos_preprocessing: Optional[str] = None,
+        add_eos_token: bool = False,
+        fuse_documents: bool = True):
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data_utils.py]:
```suggestion
        fuse_documents: bool = True) -> datasets.Dataset:
```

**Review by pablomlago — APPROVED** [review]
LGTM (assuming test pass)
